gh-127503: Emscripten make Python.sh function as proper Python CLI

hoodmane opened the pull request:
* The first commit fixes (I hope) the problems with bsd `realpath`.
* The second commit adjusts `python.sh` to mount all directories from the native file system except `/dev`, `/lib` and `/proc`.
* The third commit makes native environment variables visible in Emscripten.
* The fourth commit changes the make target to `all` so that we can run tests in-tree. Now that Python can see `pybuilddir.txt` all of the paths work correctly running in-tree.
* The fifth commit adds a NODEFLAGS environment variable that allows us to change the way we invoke `node` (for example to enable experimental features or debugging)

Fixes #127503
Fixes #127111

<!-- gh-issue-number: gh-127503 -->
* Issue: gh-127503
<!-- /gh-issue-number -->

hoodmane committed:
Fix realpath usage on systems with bsd coreutils

hoodmane committed:
Mount all native directories that aren't known to cause trouble

This mounts all native directories into the Emscripten file system except for
dev, lib, and proc.

hoodmane committed:
Pass environment variables into Emscripten

hoodmane committed:
Change make target from commoninstall to all

hoodmane committed:
Pass flags in NODEFLAGS environment variable to node

Useful for: --experimental flags, --inspect flags, etc.

hoodmane commented on Tools/wasm/emscripten/__main__.py:
```diff
@@ -215,12 +215,30 @@ def configure_emscripten_python(context, working_dir):
             # node_entry.mjs can set sys.executable correctly.
-            exec {host_runner} {node_entry} "$(realpath -s $0)" "$@"
+            # Intentionally allow word splitting on NODEFLAGS.
+            exec {host_runner} {node_entry} ${NODEFLAGS} "$($REALPATH "$0")" "$@"
```
```suggestion
            exec {host_runner} {node_entry} $NODEFLAGS "$($REALPATH "$0")" "$@"
```

mdboom commented on Tools/wasm/emscripten/__main__.py:
```diff
@@ -215,12 +215,30 @@ def configure_emscripten_python(context, working_dir):
+            """\
             #!/bin/sh
 
+            # Macs come with free BSD coreutils which doesn't have the -s option
```
```suggestion
            # Macs come with FreeBSD coreutils which doesn't have the -s option
```

mdboom commented on Tools/wasm/emscripten/__main__.py:
```diff
@@ -215,12 +215,30 @@ def configure_emscripten_python(context, working_dir):
+                REALPATH=abs_path
+            fi
+            """
+            f"""
```
This 2 strings with implicit concatenation is a bit subtle.  Can we just do one string (which probably requires replacing `{` with `{{` above).

hoodmane committed:
Apply mdboom's review comments

hoodmane committed:
Fix NODEFLAGS reference

hoodmane committed:
Fix NODEFLAGS

freakboy3742 commented on Tools/wasm/emscripten/__main__.py:
```diff
@@ -215,12 +215,29 @@ def configure_emscripten_python(context, working_dir):
             # node_entry.mjs can set sys.executable correctly.
-            exec {host_runner} {node_entry} "$(realpath -s $0)" "$@"
+            # Intentionally allow word splitting on NODEFLAGS.
+            exec node $NODEFLAGS /home/rchatham/cpython/cross-build/wasm32-emscripten/node_entry.mjs "$($REALPATH "$0")" "$@"
```
As much as I'd *like* to give you an account on my laptop... 🤣 

Should this be:
```suggestion
            exec {host_runner} $NODEFLAGS {node_entry} "$($REALPATH "$0")" "$@"
```
plus returning the script to an f-string?

hoodmane replied:
I always make this mistake with code templating: it's easier to edit the output directly until I get what I want but copying it back is tricky.

freakboy3742 commented on Tools/wasm/emscripten/__main__.py:
```diff
@@ -215,12 +215,29 @@ def configure_emscripten_python(context, working_dir):
+            if which grealpath > /dev/null; then
+                # It has brew installed gnu core utils, use that
+                REALPATH="grealpath -s"
+            elif which realpath > /dev/null && realpath --version | grep GNU > /dev/null; then
```
`realpath --version` outputs to stderr because `--version` isn't a valid option on macOS/FreeBSD
```suggestion
            elif which realpath > /dev/null && realpath --version 2&>1 | grep GNU > /dev/null; then
```

hoodmane replied:
Sorry, I already fixed this for you once and then copied over it.

freakboy3742 commented on Tools/wasm/emscripten/node_entry.mjs:
```diff
@@ -12,19 +11,34 @@ if (process?.versions?.node) {
+    Module.FS.mkdirTree("/home/");
+    mountDirectories(Module);
+    Module.FS.chdir(process.cwd());
+    Object.assign(Module.ENV, process.env);
```
For my own understanding - am I correct in saying that this means the mounted filesystem has a much more complete "traditional" filesystem structure under `/home`, rather than just a top level `/lib`?

freakboy3742 replied:
Related: Does this change require any update to the docs that landed in #127113 to reference `/home/lib`, rather than `/lib`? I guess there's nothing that *requires* the use of the `/home` prefix, but for consistency/establishing best practice, would it be advisable?

hoodmane replied:
Yeah, what I'm doing here is mounting as much as possible of your native file system. I dropped `/dev`, `/proc`, and `/lib` because they contain native devices, processes, and libraries whereas we want to tell people about the Emscripten devices and processes. Though honestly unlinking `/dev` and bringing in the native `/dev` would probably not be the worst idea. Something to think about.

hoodmane replied:
I don't think any documentation update is needed. As long as `configure.ac` sets `host_prefix` to `/lib`, mounting the standard library there will allow it to be located. I don't think it would work to mount it to `/home/lib`...

freakboy3742 reviewed (changes requested):
This works for me once I've made the couple of tweaks suggested inline; plus one question for my own edification.

hoodmane committed:
Fix python.sh template

hoodmane committed:
Be more careful with handling of the program name

Previous version only worked if the name was python.sh. If we symlink something
else to it (say `python`) it would fail. This update makes it not depend on the
name of the link.

hoodmane committed:
Apply code formatter

hoodmane committed:
Redirect realpath --version to /dev/null again

hoodmane committed:
Cleanup

hoodmane committed:
Merge branch 'main' into emscripten-cli-fixes

freakboy3742 commented on Tools/wasm/emscripten/__main__.py:
```diff
@@ -215,12 +215,29 @@ def configure_emscripten_python(context, working_dir):
     exec_script.write_text(
         dedent(
-            f"""\
+            """\
```
Are you sure this shouldn't be an f-string?
```suggestion
            f"""\
```

hoodmane replied:
Right...

hoodmane committed:
Put back in f string

freakboy3742 reviewed (approved):
👍
